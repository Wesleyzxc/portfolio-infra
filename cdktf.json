{
  "language": "typescript",
  "app": "npx ts-node main.ts",
  "projectId": "405e6ca0-fb51-4ae0-a2f3-f22c4de2ce24",
  "sendCrashReports": "false",
  "terraformProviders": [],
  "terraformModules": [],
  "context": {

  }
}
