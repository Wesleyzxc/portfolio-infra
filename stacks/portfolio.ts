import { Construct } from "constructs";
import { TerraformStack } from "cdktf";
import { AwsProvider } from "@cdktf/provider-aws/lib/provider";
import { S3Bucket } from "@cdktf/provider-aws/lib/s3-bucket";
import { CloudfrontDistribution } from "@cdktf/provider-aws/lib/cloudfront-distribution";
import { S3BucketPublicAccessBlock } from "@cdktf/provider-aws/lib/s3-bucket-public-access-block";
import { DEFAULT_REGION } from "../config";

export class Portfolio extends TerraformStack {
    private BUCKET_NAME = "wesleyzxc-portfolio";
  
    constructor(scope: Construct, id: string) {
      super(scope, id);
      
  
      new AwsProvider(this, "aws", {
        region: DEFAULT_REGION,
      });
  
      const bucket = new S3Bucket(this, "bucket", {
        bucket: this.BUCKET_NAME,
        acl: 'public-read',
        website: {indexDocument: 'index.html'},
        
        policy: `{
          "Version": "2012-10-17",
          "Statement": [
            {
              "Sid": "PublicReadGetObject",
              "Effect": "Allow",
              "Principal": "*",
              "Action": [
                "s3:GetObject"
              ],
              "Resource": [
                "arn:aws:s3:::${this.BUCKET_NAME}/*"
              ]
            }
          ]
        }`,
      });
  
      new S3BucketPublicAccessBlock(this, "bucketPublicAccess", {
        bucket: bucket.id,
  
        blockPublicAcls: false,
        blockPublicPolicy: false,
        restrictPublicBuckets: false,
        ignorePublicAcls: false,
  
        dependsOn: [bucket]
      })
  
      new CloudfrontDistribution(this, "distribution", {
        enabled: true,
        defaultRootObject: 'index.html',
        isIpv6Enabled: true,
  
        viewerCertificate: {
          cloudfrontDefaultCertificate: true,
          sslSupportMethod: "sni-only",
        },
  
        restrictions: {
          geoRestriction: {
            restrictionType: "none",
          },
        },
  
        defaultCacheBehavior: {
          allowedMethods: ['GET', 'HEAD', 'OPTIONS'],
          cachedMethods: ['GET', 'HEAD', 'OPTIONS'],
          cachePolicyId: "658327ea-f89d-4fab-a63d-7e88639e58f6",
          targetOriginId: bucket.id,
          viewerProtocolPolicy: "redirect-to-https"
        },
        origin: [{
          originId: bucket.id,
          domainName: bucket.bucketDomainName,
          customOriginConfig: {
            httpPort: 80,
            httpsPort: 443,
            originProtocolPolicy: "http-only",
            originSslProtocols: ["TLSv1.2", "TLSv1.1"],
          }
        }],
      })
    }
  }