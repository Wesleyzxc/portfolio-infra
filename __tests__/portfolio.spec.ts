import { CloudfrontDistribution } from "@cdktf/provider-aws/lib/cloudfront-distribution";
import { S3Bucket } from "@cdktf/provider-aws/lib/s3-bucket";
import { S3BucketPublicAccessBlock } from "@cdktf/provider-aws/lib/s3-bucket-public-access-block";
import { Testing } from "cdktf";
import "cdktf/lib/testing/adapters/jest";
import { Portfolio } from "../stacks/portfolio";

describe("portfolio", () => {
  it("should be a valid stack", () => {
    const app = Testing.app();
    const stack = new Portfolio(app, "my-app");

    expect(Testing.fullSynth(stack)).toBeValidTerraform();
    expect(Testing.fullSynth(stack)).toPlanSuccessfully();
  });

  it("should contain an s3 bucket with public access", () => {
    const app = Testing.app();
    const stack = new Portfolio(app, "my-app");
    const synthesized = Testing.synth(stack);

    expect(synthesized).toHaveResourceWithProperties(S3Bucket, {
      acl: "public-read",
      bucket: "wesleyzxc-portfolio",
      website: {
        index_document: "index.html"
      }
    });

    expect(synthesized).toHaveResourceWithProperties(S3BucketPublicAccessBlock, {
      block_public_acls: false,
      block_public_policy: false,
      bucket: "${aws_s3_bucket.bucket.id}",
      depends_on: [
        "aws_s3_bucket.bucket"
      ],
      ignore_public_acls: false,
      restrict_public_buckets: false
    });
  });


  it("should contain a cloudfront distribution", () => {
    const app = Testing.app();
    const stack = new Portfolio(app, "my-app");
    const synthesized = Testing.synth(stack);

    expect(synthesized).toMatchSnapshot();
  });
});