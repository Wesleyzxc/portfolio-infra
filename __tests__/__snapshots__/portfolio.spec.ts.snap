// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`portfolio should contain a cloudfront distribution 1`] = `
"{
  "provider": {
    "aws": [
      {
        "region": "ap-southeast-2"
      }
    ]
  },
  "resource": {
    "aws_cloudfront_distribution": {
      "distribution": {
        "default_cache_behavior": {
          "allowed_methods": [
            "GET",
            "HEAD",
            "OPTIONS"
          ],
          "cache_policy_id": "658327ea-f89d-4fab-a63d-7e88639e58f6",
          "cached_methods": [
            "GET",
            "HEAD",
            "OPTIONS"
          ],
          "target_origin_id": "\${aws_s3_bucket.bucket.id}",
          "viewer_protocol_policy": "redirect-to-https"
        },
        "default_root_object": "index.html",
        "enabled": true,
        "is_ipv6_enabled": true,
        "origin": [
          {
            "custom_origin_config": {
              "http_port": 80,
              "https_port": 443,
              "origin_protocol_policy": "http-only",
              "origin_ssl_protocols": [
                "TLSv1.2",
                "TLSv1.1"
              ]
            },
            "domain_name": "\${aws_s3_bucket.bucket.bucket_domain_name}",
            "origin_id": "\${aws_s3_bucket.bucket.id}"
          }
        ],
        "restrictions": {
          "geo_restriction": {
            "restriction_type": "none"
          }
        },
        "viewer_certificate": {
          "cloudfront_default_certificate": true,
          "ssl_support_method": "sni-only"
        }
      }
    },
    "aws_s3_bucket": {
      "bucket": {
        "acl": "public-read",
        "bucket": "wesleyzxc-portfolio",
        "policy": "{\\n          \\"Version\\": \\"2012-10-17\\",\\n          \\"Statement\\": [\\n            {\\n              \\"Sid\\": \\"PublicReadGetObject\\",\\n              \\"Effect\\": \\"Allow\\",\\n              \\"Principal\\": \\"*\\",\\n              \\"Action\\": [\\n                \\"s3:GetObject\\"\\n              ],\\n              \\"Resource\\": [\\n                \\"arn:aws:s3:::wesleyzxc-portfolio/*\\"\\n              ]\\n            }\\n          ]\\n        }",
        "website": {
          "index_document": "index.html"
        }
      }
    },
    "aws_s3_bucket_public_access_block": {
      "bucketPublicAccess": {
        "block_public_acls": false,
        "block_public_policy": false,
        "bucket": "\${aws_s3_bucket.bucket.id}",
        "depends_on": [
          "aws_s3_bucket.bucket"
        ],
        "ignore_public_acls": false,
        "restrict_public_buckets": false
      }
    }
  },
  "terraform": {
    "required_providers": {
      "aws": {
        "source": "aws",
        "version": "5.75.1"
      }
    }
  }
}"
`;
